app/models/manageiq/providers/openstack/inventory/collector/target_collection.rb

```diff
@@ -211,11 +213,11 @@ def cloud_volume_snapshots
   end
 
   def cloud_volume_backups
+    # backup notifications from panko don't include ids, so
+    # we will just refresh all the backups if we get an event.
     return [] if references(:cloud_volume_backups).blank?
     return @cloud_volume_backups if @cloud_volume_backups.any?
-    @cloud_volume_backups = targets_by_association(:cloud_volume_backups).collect do |target|
-      scoped_get_backup(target.manager_ref[:ems_ref], target.options[:tenant_id])
-    end.compact
+    @cloud_volume_backups = cinder_service.handled_list(:list_backups_detailed, {:__request_body_index => "backups"}, cinder_admin?)
```

Comment:
we will need specs for this, I think this way it would lead to duplication of existing cloud_volume_backups

I think what we could do to make this work on persistor could be switching ```:targeted => true/false```. Given references will be [nil] or [], then ```targeted => true``` will not update or delete anything, since the targeted query will not return any records. With ```:targeted => false```, we will do a full refresh of the cloud_volume_backups IC.

Same for the key_pairs. But lets do specs to verify this will work.

Reply:
I'll write specs to make sure this behaves correctly. Why do you suspect it would result in duplicated objects?

Reply:
so the targeted refresh takes the references and it builds a query, so it would be:

```CloudVolumeBackups.where(:ems_ref => references(...))```

So if this returns always [], the refresh will create anything built by the parser. In this case that would be all cloud_volume_backups are created again. While the correct behavior should be that the query returns the existing ones, and we just update them.

Reply:
Ah, gotcha

Reply:
Local testing as well as my new specs suggest that this duplication does not occur.

Reply:
ah, I see, the DB scope is automatically inferred from the built data :-)

but there are still 2 things left, if we don't switch off the ```:targeted``` mode:
1. The scope is just for built data, so we will not delete entities that do not exist in the provider (you can try the stubbed spec with decreasing scaling)
2. The query for full refresh is more effective than the targeted one, if we are fetching all records

Reply:
Ah, I see. `:targeted` seems to be at the persister level, not just the individual object collection, so how can I isolate one collection as not targeted while the rest are?

Reply:
Hm, you should be able to overwrite it per 1 Inventory Collection?

Reply:
It seems like if I set the key_pair collection to `:targeted => false` that it still calls TargetedCollection's key pair collection method rather than CloudManager's, and entities aren't being deleted. Here's the persister definition:

```ruby
  def add_key_pairs(extra_properties = {})
    add_collection(cloud, :key_pairs, extra_properties, :shared_properties => shared_options.merge(:targeted => false)) do |builder|
      builder.add_properties(
        :model_class => ManageIQ::Providers::Openstack::CloudManager::AuthKeyPair
      )

      builder.add_default_values(:resource => manager) unless targeted?
    end
  end
```

Here's the test that fails because objects aren't removed 
```ruby
 it "works around keypair targeted refresh by refreshing all keypairs without creating duplicates" do
  @data_scaling = 1
  setup_mocked_collector
  2.times do
    keypair_target = ManagerRefresh::Target.new(:manager     => @ems,
                                                :association => :key_pairs,
                                                :manager_ref => {:ems_ref => nil})
    EmsRefresh.refresh(keypair_target)
    expect(AuthPrivateKey.count).to eq(test_counts(@data_scaling)[:key_pairs_count])
    @data_scaling -= 1
  end
end
```
I can see that the Inventory Collection's `@targeted` property is returning `false`, so I don't think there's an issue with it being passed incorrectly. Any ideas?

Reply:
try

```ruby

 def add_key_pairs(extra_properties = {})
    add_collection(cloud, :key_pairs, extra_properties) do |builder|
      builder.add_properties(
        :model_class => ManageIQ::Providers::Openstack::CloudManager::AuthKeyPair
      )

      builder.add_properties(
        :targeted => false
      ) unless references(:key_pairs).blank?

      builder.add_default_values(:resource => manager) unless targeted?
    end
  end

```

Reply:
btw the ```builder.add_default_values(:resource => manager) unless targeted?``` looks wrong? The :resource should be filled in any refresh?

Reply:
the ```that it still calls TargetedCollection's key pair collection method rather than CloudManager's```, this should be ok right? Since you have there ```unless references(:key_pairs).blank? -> collect all keypairs```

Reply:
`builder.add_properties(:targeted => false)` doesn't make a difference. I'm not sure why `resource` is being set conditionally, I think that happened during the builder refactor. I'll investigate that later. And it's fine to call TargetedCollection's key pair method, I just was mistakenly believing I'd see a different call when `targeted => false`

Reply:
right, ```targeted => false``` should only change how we look at db. With targeted => true, we do ```ems.key_pairs.where(references(:vms))``` with false we do ``````ems.key_pairs```

So with false it loads all key_pairs and it will delete those you will not .build in parser. While with true, it just loads the ones you've built, so it will just update.

Reply:
in the spec, try also ems.key_pairs, since there might be the bug with :resource being nil

Reply:
Doesn't seem to make a difference either way, though I'm continuing to investigate.

Reply:
@Ladas Nevermind me, the spec wasn't written properly. >.>